presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java

```diff
@@ -0,0 +1,408 @@
+                throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), "Cannot handle approx_percentile function " + aggregation);
+            }
+
+            String fractionString;
+            RowExpression fractionInput = inputs.get(1);
+
+            if (fractionInput instanceof ConstantExpression) {
+                fractionString = PinotProjectExpressionConverter.getLiteralAsString((ConstantExpression) fractionInput);
+            }
+            else if (fractionInput instanceof VariableReferenceExpression) {
+                Selection fraction = inputSelections.get(fractionInput);
+                if (fraction.getOrigin() != LITERAL) {
+                    throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(),
+                            "Cannot handle approx_percentile percentage argument be a non literal " + aggregation);
+                }
+                fractionString = fraction.getDefinition();
+            }
+            else {
+                throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), "Expected the fraction to be a constant or a variable " + fractionInput);
+            }
+            int percentile = getValidPercentile(fractionString);
+            if (percentile < 0) {
+                throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(),
+                        format("Cannot handle approx_percentile parsed as %d from input %s (function %s)", percentile, fractionString, aggregation));
+            }
+            return format("PERCENTILEEST%d(%s)", percentile, inputSelections.get(getVariableReference(inputs.get(0))));
+        }
+
+        private int getValidPercentile(String fraction)
+        {
+            try {
+                double p = Double.parseDouble(fraction);
+                if (p < 0 || p > 1) {
+                    throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Percentile must be between 0 and 1");
+                }
+                double percent = p * 100.0;
+                if (percent == Math.floor(percent)) {
+                    return (int) percent;
+                }
+            }
+            catch (NumberFormatException ne) {
+                // Skip
+            }
+            return -1;
+        }
+
+        @Override
+        public PinotQueryGeneratorContext visitAggregation(AggregationNode node, PinotQueryGeneratorContext context)
+        {
+            context = node.getSource().accept(this, context);
+            requireNonNull(context, "context is null");
+            checkSupported(!node.getStep().isOutputPartial(), "partial aggregations are not supported in Pinot pushdown framework");
+            checkSupported(preferBrokerQueries, "Cannot push aggregation in segment mode");
+
+            LinkedHashMap<VariableReferenceExpression, Selection> newSelections = new LinkedHashMap<>();
+            LinkedHashSet<VariableReferenceExpression> groupByColumns = new LinkedHashSet<>();
```

Comment:
Set<...>

Reply:
I very very much want this to be order sensitive and thus would like to let this remain a LinkedHashMap, LinkedHashSet etc. We rely on this insertion order when generating the pinot query. Thus the "PinotQueryGeneratorContext.withXXX" require a linkedhashset/map.